server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -429,31 +528,130 @@ public int hashCode() {
     synchronized void cleanCache() {
         final Set<CleanupKey> currentKeysToClean = new HashSet<>();
         final Set<Object> currentFullClean = new HashSet<>();
-        currentKeysToClean.clear();
-        currentFullClean.clear();
-        for (Iterator<CleanupKey> iterator = keysToClean.iterator(); iterator.hasNext();) {
-            CleanupKey cleanupKey = iterator.next();
-            iterator.remove();
-            if (cleanupKey.readerCacheKeyId == null || cleanupKey.entity.isOpen() == false) {
-                // null indicates full cleanup, as does a closed shard
+        /*
+        Stores the keys that need to be removed from keysToClean
+        This is done to avoid ConcurrentModificationException
+        */
+        final Set<CleanupKey> keysCleanedFromAllCaches = new HashSet<>();
+
+        for (Map.Entry<CleanupKey, CleanupStatus> entry : keysToClean.entrySet()) {
+            CleanupKey cleanupKey = entry.getKey();
+            CleanupStatus cleanupStatus = entry.getValue();
+
+            if (cleanupStatus.cleanedInHeap && cleanupStatus.cleanedOnDisk) {
+                keysCleanedFromAllCaches.add(cleanupKey);
+                continue;
+            }
+
+            if (cleanupStatus.cleanedInHeap) continue;
+
+            if (needsFullClean(cleanupKey)) {
                 currentFullClean.add(cleanupKey.entity.getCacheIdentity());
             } else {
                 currentKeysToClean.add(cleanupKey);
             }
+            cleanupStatus.cleanedInHeap = true;
         }
-        if (!currentKeysToClean.isEmpty() || !currentFullClean.isEmpty()) {
-            for (Iterator<Key> iterator = tieredCacheService.getOnHeapCachingTier().keys().iterator(); iterator.hasNext();) {
-                Key key = iterator.next();
-                if (currentFullClean.contains(key.entity.getCacheIdentity())) {
-                    iterator.remove();
+
+        // Remove keys that have been cleaned from all caches
+        keysToClean.keySet().removeAll(keysCleanedFromAllCaches);
+
+        // Early exit if no cleanup is needed
+        if (currentKeysToClean.isEmpty() && currentFullClean.isEmpty()) {
+            return;
+        }
+
+        cleanUpKeys(
+            tieredCacheService.getOnHeapCachingTier(),
+            currentKeysToClean,
+            currentFullClean
+        );
+        tieredCacheService.getOnHeapCachingTier().refresh();
+    }
+
+    /**
+     * Logic to clean up disk based cache.
+     * <p>
+     * TODO: cleanDiskCache is very specific to disk caching tier. We can refactor this to be more generic.
+     */
+    synchronized void cleanDiskCache(double diskCachesCleanThresholdPercent) {
```

Comment:
Besides the check for disk cleanup key percentage, which cleanupStatus field gets set to true, and the tier to call cleanupKeys on in the end, this is pretty much the same function as cleanCache. Could we refactor to reuse that code?